File: android/framework_old/src/main/java/com/dreampany/nearby/misc/SmartQueue.java
package com.dreampany.nearby.misc;

import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.TimeUnit;

/**
 * Created by Hawladar Roman on 6/4/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
public class SmartQueue<T> extends LinkedBlockingDeque<T> {

    public boolean insertFirst(T t) {
        if (t == null) {
            return false;
        }
        while (true) {
            try {
                putFirst(t);
                return false;
            } catch (InterruptedException e) {
            }
        }
    }

    public boolean insertFirstUniquely(T t) {
        if (t == null) {
            return false;
        }
        if (contains(t)) {
            remove(t);
        }
        while (true) {
            try {
                putFirst(t);
                return false;
            } catch (InterruptedException e) {
            }
        }
    }

    public boolean insertLast(T t) {
        if (t == null) {
            return false;
        }
        while (true) {
            try {
                super.putLast(t);
                return false;
            } catch (InterruptedException e) {
            }
        }
    }

    public boolean insertLastUniquely(T t) {
        if (t == null) {
            return false;
        }
        if (contains(t)) {
            remove(t);
        }
        while (true) {
            try {
                super.putLast(t);
                return false;
            } catch (InterruptedException e) {
            }
        }
    }

    public T pollFirst() {
        return super.pollFirst();
    }

    public T peekFirst() {
        return super.peekFirst();
    }

    public T pollFirst(long timeout) {
        try {
            return super.pollFirst(timeout, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            return null;
        }
    }

    public T takeFirst() {
        try {
            return super.takeFirst();
        } catch (InterruptedException e) {
            return null;
        }
    }

    @Override
    public boolean remove(Object o) {
        return super.remove(o);
    }

    public void removeAll() {
        super.clear();
    }
}

File: android/network/src/main/java/com/dreampany/network/misc/SmartQueue.kt
package com.dreampany.network.misc

import java.util.concurrent.LinkedBlockingDeque
import java.util.concurrent.TimeUnit

/**
 * Created by roman on 7/2/21
 * Copyright (c) 2021 butler. All rights reserved.
 * ifte.net@gmail.com
 * Last modified $file.lastModified
 */
class SmartQueue<T> : LinkedBlockingDeque<T>() {

    fun insertFirst(t: T?): Boolean {
        if (t == null) {
            return false
        }
        while (true) {
            try {
                putFirst(t)
                return false
            } catch (ignored: InterruptedException) {
            }
        }
    }

    fun insertFirstUniquely(t: T?): Boolean {
        if (t == null) {
            return false
        }
        if (contains(t)) {
            remove(t)
        }
        while (true) {
            try {
                putFirst(t)
                return false
            } catch (ignored: InterruptedException) {
            }
        }
    }

    fun insertLast(t: T?): Boolean {
        if (t == null) {
            return false
        }
        while (true) {
            try {
                super.putLast(t)
                return false
            } catch (ignored: InterruptedException) {
            }
        }
    }

    fun insertLastUniquely(t: T?): Boolean {
        if (t == null) {
            return false
        }
        if (contains(t)) {
            remove(t)
        }
        while (true) {
            try {
                super.putLast(t)
                return false
            } catch (ignored: InterruptedException) {
            }
        }
    }

    fun pollFirst(timeout: Long): T? {
        return try {
            super.pollFirst(timeout, TimeUnit.MILLISECONDS)
        } catch (ignored: InterruptedException) {
            null
        }
    }

    override fun takeFirst(): T? {
        return try {
            super.takeFirst()
        } catch (ignored: InterruptedException) {
            null
        }
    }

    override fun pollFirst(): T? {
        return try {
            super.pollFirst()
        } catch (ignored: InterruptedException) {
            null
        }
    }
}